README.md
<a href="https://www.gotoiot.com/">
    <img src="doc/gotoiot-logo.png" alt="logo" title="Goto IoT" align="right" width="60" height="60" />
</a>

Web App Full Stack Base
=======================

*Ayudaría mucho si apoyaras este proyecto con una ⭐ en Github!*

Este proyecto es una aplicación web fullstack que se ejecuta sobre el ecosistema `Docker`. Está compuesta por un compilador de `TypeScript` que te permite utilizar este superset de JavaScript para poder programar un `cliente web`. También tiene un servicio en `NodeJS` que te permite ejecutar código en backend y al mismo tiempo disponibilizar el código del cliente web para interactar con el servicio. Además tiene una `base de datos` MySQL que puede interactuar con el backend para guardar y consultar datos, y de manera adicional trae un `administrador` de base de datos para poder administrar la base en caso que lo necesites.

La aplicación IoT de base que viene con este proyecto se encarga de crear una tabla llamada `Devices` en la base de datos, y la idea es que vos puedas desarrollar el código de backend y frontend que te permita controlar desde el navegador el estado de los devices de un hogar inteligente - *como pueden ser luces, TVs, ventiladores, persianas, enchufes y otros* - y almacenar los estados de cada uno en la base de datos. 

Realizando estas tareas vas a a tener una aplicación fullstack IoT del mundo real que utiliza tecnologías actuales en la que un backend es capaz de interactuar con una DB para cumplir con las peticiones de control que se le mandan desde el cliente web.

En esta imagen podés ver una posible implementación del cliente web que controla los artefactos del hogar.

![architecture](doc/webapp-example-1.png)

## Comenzando 🚀

Esta sección es una guía con los pasos escenciales para que puedas poner en marcha la aplicación.

<details><summary><b>Mira los pasos necesarios</b></summary><br>

### Instalar las dependencias

Para correr este proyecto es necesario que instales `Docker` y `Docker Compose`. 

En [este artículo](https://www.gotoiot.com/pages/articles/docker_installation_linux/) publicado en nuestra web están los detalles para instalar Docker y Docker Compose en una máquina Linux. Si querés instalar ambas herramientas en una Raspberry Pi podés seguir [este artículo](https://www.gotoiot.com/pages/articles/rpi_docker_installation) de nuestra web que te muestra todos los pasos necesarios.

En caso que quieras instalar las herramientas en otra plataforma o tengas algún incoveniente, podes leer la documentación oficial de [Docker](https://docs.docker.com/get-docker/) y también la de [Docker Compose](https://docs.docker.com/compose/install/).

Continua con la descarga del código cuando tengas las dependencias instaladas y funcionando.

### Descargar el código

Para descargar el código, lo más conveniente es que realices un `fork` de este proyecto a tu cuenta personal haciendo click en [este link](https://github.com/gotoiot/app-fullstack-base/fork). Una vez que ya tengas el fork a tu cuenta, descargalo con este comando (acordate de poner tu usuario en el link):

```
git clone https://github.com/USER/app-fullstack-base.git
```

> En caso que no tengas una cuenta en Github podes clonar directamente este repo.

### Ejecutar la aplicación

Para ejecutar la aplicación tenes que correr el comando `docker-compose up` desde la raíz del proyecto. Este comando va a descargar las imágenes de Docker de node, de typescript, de la base datos y del admin de la DB, y luego ponerlas en funcionamiento. 

Para acceder al cliente web ingresa a a la URL [http://localhost:8000/](http://localhost:8000/) y para acceder al admin de la DB accedé a [localhost:8001/](http://localhost:8001/). 

Si pudiste acceder al cliente web y al administrador significa que la aplicación se encuentra corriendo bien. 

> Si te aparece un error la primera vez que corres la app, deteńe el proceso y volvé a iniciarla. Esto es debido a que el backend espera que la DB esté creada al iniciar, y en la primera ejecución puede no alcanzar a crearse. A partir de la segunda vez el problema queda solucionado.

</details>

Continuá explorando el proyecto una vez que lo tengas funcionando.

## Configuraciones de funcionamiento 🔩

Al crearse la aplicación se ejecutan los contenedores de Docker de cada servicio, se crea la base de datos y sus tablas. A continuación podés encontrar info si querés cambiar la estructura de la DB o bien sus configuraciones de acceso.

<details><summary><b>Lee cómo configurar la aplicación</b></summary><br>

### Configuración de la DB

Como ya comprobaste, para acceder PHPMyAdmin tenés que ingresar en la URL [localhost:8001/](http://localhost:8001/). En el login del administrador, el usuario para acceder a la db es `root` y contraseña es la variable `MYSQL_ROOT_PASSWORD` del archivo `docker-compose.yml`.

Para el caso del servicio de NodeJS que se comunica con la DB fijate que en el archivo `src/backend/mysql-connector.js` están los datos de acceso para ingresar a la base.

Si quisieras cambiar la contraseña, puertos, hostname u otras configuraciones de la DB deberías primero modificar el servicio de la DB en el archivo `docker-compose.yml` y luego actualizar las configuraciones para acceder desde PHPMyAdmin y el servicio de NodeJS.

### Estructura de la DB

Al iniciar el servicio de la base de datos, si esta no está creada toma el archivo que se encuentra en `db/dumps/smart_home.sql` para crear la base de datos automáticamente.

En ese archivo está la configuración de la tabla `Devices` y otras configuraciones más. Si quisieras cambiar algunas configuraciones deberías modificar este archivo y crear nuevamente la base de datos para que se tomen en cuenta los cambios.

Tené en cuenta que la base de datos se crea con permisos de superusuario por lo que no podrías borrar el directorio con tu usuario de sistema, para eso debés hacerlo con permisos de administrador. En ese caso podés ejecutar el comando `sudo rm -r db/data` para borrar el directorio completo.

</details>


## Detalles principales 🔍

En esta sección vas a encontrar las características más relevantes del proyecto.

<details><summary><b>Mira los detalles más importantes de la aplicación</b></summary><br>
<br>

### Arquitectura de la aplicación

Como ya pudiste ver, la aplicación se ejecuta sobre el ecosistema Docker, y en esta imagen podés ver el diagrama de arquitectura.

![architecture](doc/architecture.png)

### El cliente web

El cliente web es una Single Page Application que se comunica con el servicio en NodeJS mediante JSON a través de requests HTTP. Puede consultar el estado de dispositivos en la base de datos (por medio del servicio en NodeJS) y también cambiar el estado de los mismos. Los estilos del código están basados en **Material Design**.

### El servicio web

El servicio en **NodeJS** posee distintos endpoints para comunicarse con el cliente web mediante requests HTTP enviando **JSON** en cada transacción. Procesando estos requests es capaz de comunicarse con la base de datos para consultar y controlar el estado de los dispositivos, y devolverle una respuesta al cliente web también en formato JSON. Así mismo el servicio es capaz de servir el código del cliente web.

### La base de datos

La base de datos se comunica con el servicio de NodeJS y permite almacenar el estado de los dispositivos en la tabla **Devices**. Ejecuta un motor **MySQL versión 5.7** y permite que la comunicación con sus clientes pueda realizarse usando usuario y contraseña en texto plano. En versiones posteriores es necesario brindar claves de acceso, por este motivo la versión 5.7 es bastante utilizada para fases de desarrollo.

### El administrador de la DB

Para esta aplicación se usa **PHPMyAdmin**, que es un administrador de base de datos web muy utilizado y que podés utilizar en caso que quieras realizar operaciones con la base, como crear tablas, modificar columnas, hacer consultas y otras cosas más.

### El compilador de TypeScript

**TypeScript** es un lenguaje de programación libre y de código abierto desarrollado y mantenido por Microsoft. Es un superconjunto de JavaScript, que esencialmente añade tipos estáticos y objetos basados en clases. Para esta aplicación se usa un compilador de TypeScript basado en una imagen de [Harmish](https://hub.docker.com/r/harmish) en Dockerhub, y está configurado para monitorear en tiempo real los cambios que se realizan sobre el directorio **src/frontend/ts** y automáticamente generar código compilado a JavaScript en el directorio  **src/frontend/js**. Los mensajes del compilador aparecen automáticamente en la terminal al ejecutar el comando **docker-compose up**.

### Ejecución de servicios

Los servicios de la aplicación se ejecutan sobre **contenedores de Docker**, así se pueden desplegar de igual manera en diferentes plataformas. Los detalles sobre cómo funcionan los servicios los podés ver directamente en el archivo **docker-compose.yml**.

### Organización del proyecto

En la siguiente ilustración podés ver cómo está organizado el proyecto para que tengas en claro qué cosas hay en cada lugar.

```sh
├── db                          # directorio de la DB
│   ├── data                    # estructura y datos de la DB
│   └── dumps                   # directorio de estructuras de la DB
│       └── smart_home.sql      # estructura con la base de datos "smart_home"
├── doc                         # documentacion general del proyecto
└── src                         # directorio codigo fuente
│   ├── backend                 # directorio para el backend de la aplicacion
│   │   ├── index.js            # codigo principal del backend
│   │   ├── mysql-connector.js  # codigo de conexion a la base de datos
│   │   ├── package.json        # configuracion de proyecto NodeJS
│   │   └── package-lock.json   # configuracion de proyecto NodeJS
│   └── frontend                # directorio para el frontend de la aplicacion
│       ├── js                  # codigo javascript que se compila automáticamente
│       ├── static              # donde alojan archivos de estilos, imagenes, fuentes, etc.
│       ├── ts                  # donde se encuentra el codigo TypeScript a desarrollar
│       └── index.html          # archivo principal del cliente HTML
├── docker-compose.yml          # archivo donde se aloja la configuracion completa
├── README.md                   # este archivo
├── CHANGELOG.md                # archivo para guardar los cambios del proyecto
├── LICENSE.md                  # licencia del proyecto
```

> No olvides ir poniendo tus cambios en el archivo `CHANGELOG.md` a medida que avanzas en el proyecto.

</details>

## Detalles de implementación 💻

En esta sección podés ver los detalles específicos de funcionamiento del código y que son los siguientes.

<details><summary><b>Mira los detalles de implementación</b></summary><br>

### Agregar un dispositivo

En primer lugar, al acceder al frontend, vas a encontrar un botón en la parte superior del listado de dispositivos:

![home](doc/home.png)

Al seleccionarlo, se va a abrir un menú (modal) solicitando los siguientes datos:
 - nombre
 - descripción
 - tipo: puede ser del tipo booleano (ON/OFF), valores discretos (0-1-2-3) o valores dimmerizables (0.00 a 1.00)

### Frontend

El frontend se compone un componente principal denominado <b>Main</b> que se encarga de:
  1. inicializar dispositivos (función <i>consultarDispositivosAlServidor</i>)
  2. configurar <i>event listener</i> para los principales eventos (son 5, en el <b>main.ts</b> se encuentra la función)
  3. atender diferentes eventos de acciones en la webapp

Además, el componente principal utiliza otro componente (<b>Framework</b>) para delegar todas las peticiones (<i>GET,POST,PUT,DELETE, etc.</i>) al backend. Este componente además, gestiona algunas características visuales como las alertas o mensajes emergentes.

Para editar los dispositivos, se atiende el evento de edición y se setean las variables del dispositivo en el modal de edición, junto con su ID para posterior guardado en el backend.


### Backend

El backend se encuentra conformado por un <i>index.js</i> que contiene todas las funcionalidades requeridas:
  1. obtener todos los dispositivos
  2. obtener un dispositivo
  3. cargar un dispositivo
  4. actualizar el estado de un dispositivo
  5. modificar los datos de un dispositivo
  6. eliminar un dispositivo

Además, se encuentran disponibles algunas validaciones en el archivo <i>utils.js</i> que facilitan la verificación de los datos recibidos por los endpoints (ver si los datos del dispositivo son válidos, ver si el estado es válido, etc.).

Con respecto a las validaciones, cada endpoint valida y devuelve los estados correspondientes (junto a un mensaje descriptivo) como:
  - 404 si el dispositivo no existe
  - 400 si la información no es valida
  - 500 si hubo otro error interno
  - 200 si se pudo llevar a cabo la operación

Por ejemplo, si se actualiza el estado de un dispositivo del tipo 0 (ON/OFF) el estado no puede ser 0.5, debe ser 0 o 1 solamente.
Otro caso de ejemplo, si se actualiza el nombre y/o descripción de un dispositivo (la actualización de datos puede ser total o parcial), los campos no pueden estar vacíos.

Por último, todo se encuentra almacenado en una base de datos MySQL persistente.

<details><summary><b>Ver los endpoints disponibles</b></summary><br>

Completá todos los endpoints del backend con los metodos disponibles, los headers y body que recibe, lo que devuelve, ejemplos, etc.

1) Devolver el estado de los dispositivos (<b><i>/devices</i></b>)

```json
{
    "method": "get",
    "request_headers": "application/json",
    "request_body": "",
    "response_code": 200,
    "request_body":
        [
            {
                "id": 1,
                "status": 0,
                "description": "Kitchen light",
                "name": "Device 1",
                "type": 1
            }
        ]
}
``` 

2) Devolver el estado de un dispositivo (<b><i>/device/\<id\></i></b>)

```json
{
    "method": "get",
    "request_headers": "application/json",
    "request_body": "",
    "response_code": 200,
    "request_body":
            {
                "id": 1,
                "status": 0,
                "description": "Kitchen light",
                "name": "Device 1",
                "type": 1
            }
}
``` 

3) Crear un dispositivo (<b><i>/device</i></b>)

```json
{
    "method": "post",
    "request_headers": "application/json",
    "request_body": 
        {
            "name": "Device 1",
            "description": "Kitchen light",
            "type": 1
        },
    "response_code": 200,
    "request_body": ""
}
``` 

4) Actualizar el estado de un dispositivo (<b><i>/device/\<id\>/state</i></b>)

```json
{
    "method": "post",
    "request_headers": "application/json",
    "request_body": 
        {
            "state": 0,
        },
    "response_code": 200,
    "request_body": ""
}
``` 

5) Borrar un dispositivo (<b><i>/device/\<id\></i></b>)

```json
{
    "method": "delete",
    "request_headers": "application/json",
    "request_body": "",
    "response_code": 200,
    "request_body": ""
}
``` 

6) Actualizar un dispositivo (<b><i>/device/\<id\></i></b>)

```json
{
    "method": "put",
    "request_headers": "application/json",
    "request_body": "",
    "response_code": 200,
    "request_body": ""
}
``` 

</details>

</details>

## Tecnologías utilizadas 🛠️

En esta sección podés ver las tecnologías más importantes utilizadas.

<details><summary><b>Mira la lista completa de tecnologías</b></summary><br>

* [Docker](https://www.docker.com/) - Ecosistema que permite la ejecución de contenedores de software.
* [Docker Compose](https://docs.docker.com/compose/) - Herramienta que permite administrar múltiples contenedores de Docker.
* [Node JS](https://nodejs.org/es/) - Motor de ejecución de código JavaScript en backend.
* [MySQL](https://www.mysql.com/) - Base de datos para consultar y almacenar datos.
* [PHPMyAdmin](https://www.phpmyadmin.net/) - Administrador web de base de datos.
* [Material Design](https://material.io/design) - Bibliotecas de estilo responsive para aplicaciones web.
* [TypeScript](https://www.typescriptlang.org/) - Superset de JavaScript tipado y con clases.

</details>

## Contribuir 🖇️

Si estás interesado en el proyecto y te gustaría sumar fuerzas para que siga creciendo y mejorando, podés abrir un hilo de discusión para charlar tus propuestas en [este link](https://github.com/gotoiot/app-fullstack-base/issues/new). Así mismo podés leer el archivo [Contribuir.md](https://github.com/gotoiot/gotoiot-doc/wiki/Contribuir) de nuestra Wiki donde están bien explicados los pasos para que puedas enviarnos pull requests.

## Sobre Goto IoT 📖

Goto IoT es una plataforma que publica material y proyectos de código abierto bien documentados junto a una comunidad libre que colabora y promueve el conocimiento sobre IoT entre sus miembros. Acá podés ver los links más importantes:

* **[Sitio web](https://www.gotoiot.com/):** Donde se publican los artículos y proyectos sobre IoT. 
* **[Github de Goto IoT:](https://github.com/gotoiot)** Donde están alojados los proyectos para descargar y utilizar. 
* **[Comunidad de Goto IoT:](https://groups.google.com/g/gotoiot)** Donde los miembros de la comunidad intercambian información e ideas, realizan consultas, solucionan problemas y comparten novedades.
* **[Twitter de Goto IoT:](https://twitter.com/gotoiot)** Donde se publican las novedades del sitio y temas relacionados con IoT.
* **[Wiki de Goto IoT:](https://github.com/gotoiot/doc/wiki)** Donde hay información de desarrollo complementaria para ampliar el contexto.

## Muestas de agradecimiento 🎁

Si te gustó este proyecto y quisieras apoyarlo, cualquiera de estas acciones estaría más que bien para nosotros:

* Apoyar este proyecto con una ⭐ en Github para llegar a más personas.
* Sumarte a [nuestra comunidad](https://groups.google.com/g/gotoiot) abierta y dejar un feedback sobre qué te pareció el proyecto.
* [Seguirnos en twitter](https://github.com/gotoiot/doc/wiki) y dejar algún comentario o like.
* Compartir este proyecto con otras personas.

## Autores 👥

Las colaboraciones principales fueron realizadas por:

* **[Agustin Bassi](https://github.com/agustinBassi)**: Ideación, puesta en marcha y mantenimiento del proyecto.
* **[Ernesto Giggliotti](https://github.com/ernesto-g)**: Creación inicial del frontend, elección de Material Design.
* **[Brian Ducca](https://github.com/brianducca)**: Ayuda para conectar el backend a la base de datos, puesta a punto de imagen de Docker.

También podés mirar todas las personas que han participado en la [lista completa de contribuyentes](https://github.com/###/contributors).

## Licencia 📄

Este proyecto está bajo Licencia ([MIT](https://choosealicense.com/licenses/mit/)). Podés ver el archivo [LICENSE.md](LICENSE.md) para más detalles sobre el uso de este material.

---

**Copyright © Goto IoT 2021** ⌨️ [**Website**](https://www.gotoiot.com) ⌨️ [**Group**](https://groups.google.com/g/gotoiot) ⌨️ [**Github**](https://www.github.com/gotoiot) ⌨️ [**Twitter**](https://www.twitter.com/gotoiot) ⌨️ [**Wiki**](https://github.com/gotoiot/doc/wiki)


CHANGELOG.md
# DAW Base App - Changes Log

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/) and this project adheres to [Semantic Versioning](http://semver.org/).

## 3.0.0

* Project modification
    * Introduce three device type with endpoints and validation
    * Implements and refactors device management
    * Implements frontend styling
    * Adds new info to README accordingly

## 2.2.0

* Project modification
    * Adds TypeScript compiler service to Docker Compose
    * Reestructures frontend folder for TypeScript
    * Adds new info to README accordingly
    * Changes project architecture image

## 2.1.0

* Project modification
    * Enhaces README accordingly to Goto IoT
    * Adds example of finished application
    * Removes unnecessary frontend images
    * Changes src code folders names

## 2.0.0

* Project modification
    * Changes project and organization names
    * Removes Typescript container
    * Removes Typescript Code
    * Executes Javascript code directly
    * Changes licence to MIT
    * Modifies README accordingly

## 1.0.0

* Project creation
    * Docker Compose implementation for whole project.
    * Typescript compilation into docker-compose.
    * MySQL 5.7 DB Server.
    * PHPMyAdmin.
    * NodeJS backend application.
    * Materialize CSS framework.


src/backend/utils.js
module.exports = {

    // Check if new device has name, description and type
    deviceIsValid: function (device) {
        return 'name' in device && 'description' in device && 'type' in device;
    },

    // Check if device new data is valid 
    deviceDataIsValid: function (device) {
        var validName = (!('name' in device)) || ('name' in device && device.name != "");
        var validDescription = (!('description' in device)) || ('description' in device && device.description != "");
        var validType = (!('type' in device)) || ('type' in device && [0, 1, 2].includes(device.type));

        return validName && validDescription && validType;
    },

    // Check if new device state is valid
    // If type is 0 (switch), state can only be 0 or 1 like a light with ON/OFF toggle
    // If type is 1 (discrete), state can have only step value (0, 1, 2, 3, etc.) like speed of a fan
    // If type is 2 (dimmable), state can be a float number between 0 and 1 like a dimmable light
    deviceStateIsValid: function (device, state) {
        return (device.type == 0 && [0, 1].includes(state))
            || (device.type == 1 && Number.isInteger(state) && state >= 0)
            || (device.type == 2 && state >= 0 && state <= 1);
    },
};

src/frontend/ts/framework.ts
class Framework {
  backendURL = "http://localhost:8000";

  public getDevices(responseHandler: HandleResponse) {
    let xmlHttp = new XMLHttpRequest();

    xmlHttp.onreadystatechange = () => {
      if (xmlHttp.readyState == 4) {
        if (xmlHttp.status == 200) {
          let listaDisp: Array<Device> = JSON.parse(xmlHttp.responseText);
          responseHandler.cargarGrilla(listaDisp);
        } else {
          this.mostrarAlerta("Hubo un error al comunicarse con el servidor: " + xmlHttp.status);
        }
      }
    };
    xmlHttp.open("GET", `${this.backendURL}/devices`, true);
    xmlHttp.setRequestHeader("Content-Type", "application/json");
    xmlHttp.send();
  }

  public updateDeviceState(id: number, data: number) {
    let xmlHttp = new XMLHttpRequest();
    let state = { state: data };
    xmlHttp.onreadystatechange = () => {
      if (xmlHttp.readyState == 4) {
        if (xmlHttp.status == 200) {
          this.mostrarAlerta(`Dispositivo ${id} actualizado`);
          return;
        } else {
          this.mostrarAlerta("Hubo un error al comunicarse con el servidor:" + xmlHttp.status);
        }
      }
    };
    xmlHttp.open("POST", `${this.backendURL}/device/${id}/state`, true);
    xmlHttp.setRequestHeader("Content-Type", "application/json");
    xmlHttp.send(JSON.stringify(state));
  }

  public updateDevice(
    id: number,
    device: Device,
    responseHandler: HandleResponse
  ) {
    let xmlHttp = new XMLHttpRequest();
    xmlHttp.onreadystatechange = () => {
      if (xmlHttp.readyState == 4) {
        if (xmlHttp.status == 200) {
          this.mostrarAlerta(`Dispositivo ${id} modificado`);
          responseHandler.clearNewDeviceForm();
          return;
        } else {
          this.mostrarAlerta("Hubo un error al comunicarse con el servidor:" + xmlHttp.status);
        }
      }
    };
    xmlHttp.open("PUT", `${this.backendURL}/device/${id}`, true);
    xmlHttp.setRequestHeader("Content-Type", "application/json");
    xmlHttp.send(JSON.stringify(device));
  }

  public deleteDevice(id: number) {
    let xmlHttp = new XMLHttpRequest();
    xmlHttp.onreadystatechange = () => {
      if (xmlHttp.readyState == 4) {
        if (xmlHttp.status == 200) {
          this.mostrarAlerta(`Dispositivo ${id} eliminado`);
          return;
        } else {
          this.mostrarAlerta("Hubo un error al comunicarse con el servidor: " + xmlHttp.status);
        }
      }
    };
    xmlHttp.open("DELETE", `${this.backendURL}/device/${id}`, true);
    xmlHttp.send();
  }

  public createDevice(name: string, description: string, type: number, responseHandler: HandleResponse) {
    let xmlHttp = new XMLHttpRequest();
    let device = new Device(name, description, type);
    xmlHttp.onreadystatechange = () => {
      if (xmlHttp.readyState == 4) {
        if (xmlHttp.status == 200) {
          this.mostrarAlerta(`Dispositivo creado`);
          responseHandler.clearNewDeviceForm();
          return;
        } else {
          this.mostrarAlerta("Hubo un error al comunicarse con el servidor: " + xmlHttp.status);
        }
      }
    };
    xmlHttp.open("POST", `${this.backendURL}/device`, true);
    xmlHttp.setRequestHeader("Content-Type", "application/json");
    xmlHttp.send(JSON.stringify(device));
  }

  public mostrarAlerta(message: string) {
    M.toast({ html: message });
  }

  public mostrarCargando() {
    let imgLoading = document.getElementById("loading");
    imgLoading.hidden = false;
  }

  public ocultarCargando() {
    let imgLoading = document.getElementById("loading");
    imgLoading.hidden = true;
  }
}


src/backend/index.js
//=======[ Settings, Imports & Data ]==========================================

var PORT = 3000;

var express = require('express');
const fs = require('fs');
var path = require('path');
var app = express();
var utils = require('./mysql-connector');
const { deviceIsValid, deviceDataIsValid, deviceStateIsValid } = require('./utils');

// We won't be using this because we have MySQL data
var ephimeralDevices = [];

fs.readFile(path.join(__dirname, 'datos.json'), 'utf8', (err, data) => {
    if (err) {
        console.log("Error loading data: " + err);
        return;
    }
    ephimeralDevices = JSON.parse(data);
    console.log("Initial data read");
});

// to parse application/json
app.use(express.json());
// to serve static files
app.use(express.static('/home/node/app/static/'));

//=======[ Main module code ]==================================================

app.get('/devices', function (req, res) {

    utils.query('SELECT id, name, description, state, type FROM `Devices`',
        function (error, results, fields) {
            if (error) {
                console.log(error);
                res.status(500).send({ state: "An internal server error ocurred" });
            }
            else {
                res.status(200).send(results);
            }
        });


});

app.get('/device/:id', function (req, res) {

    utils.query('SELECT id, name, description, state, type FROM `Devices` where id = ?', [req.params.id],
        function (error, results, fields) {
            if (error) {
                console.log(error);
                res.status(500).send({ state: "An internal server error ocurred" });
            }
            else {
                if (results.length > 0)
                    res.status(200).json(results[0]);
                else
                    res.sendStatus(404);
            }
        });

});

app.post('/device', function (req, res) {
    if (deviceIsValid(req.body)) {
        // We initialize device state as 0
        utils.query('INSERT INTO `Devices`(name, description, type, state) VALUES (?, ?, ?, ?)', [req.body.name, req.body.description, req.body.type, 0],
            function (error, results, fields) {
                if (error) {
                    console.log(error);
                    res.status(500).send({ state: "An internal server error ocurred" });
                }
                else {
                    res.sendStatus(200);
                }
            });
    }
    else {
        res.status(400).send("Invalid device, attributes are missing");
    }

});

app.post('/device/:id/state', function (req, res) {
    utils.query('SELECT id, type FROM `Devices` where id = ?', [req.params.id],
        function (error, results, fields) {
            if (error) {
                console.log(error);
                res.send({ state: "An internal server error ocurred" }).status(500);
            }
            else {
                if (results.length > 0) {
                    if ('state' in req.body) {
                        var currentDevice = results[0];
                        if (deviceStateIsValid(currentDevice, req.body.state)) {
                            utils.query('UPDATE `Devices` set state= ? where id= ?', [req.body.state, currentDevice.id],
                                function (error, results, fields) {
                                    if (error) {
                                        console.log(error);
                                        res.status(500).send({ state: "An internal server error ocurred while updating device state" });
                                    }
                                    else {
                                        res.sendStatus(200);
                                    }
                                });
                        }
                        else {
                            res.status(400).send({ error: "State is not valid for this device." });
                        }
                    }
                    else {
                        res.status(400).send({ error: "Invalid format when sending state: body should be a JSON with 'state' attribute." });
                    }

                }
                else
                    res.sendStatus(404);
            }
        });
});

app.delete('/device/:id', function (req, res) {
    utils.query('SELECT id, type FROM `Devices` where id = ?', [req.params.id],
        function (error, results, fields) {
            if (error) {
                console.log(error);
                res.send({ state: "An internal server error ocurred" }).status(500);
            }
            else {
                if (results.length > 0) {
                    utils.query('DELETE FROM `Devices`  where id= ?', [req.params.id],
                        function (error, results, fields) {
                            if (error) {
                                console.log(error);
                                res.status(500).send({ state: "An internal server error ocurred while deleting device" });
                            }
                            else {
                                res.sendStatus(200);
                            }
                        });
                }
                else
                    res.sendStatus(404);
            }
        });
});

// We use PUT but we allow partial updates
app.put('/device/:id', function (req, res) {
    try {
        utils.query('SELECT id, name, description, type, state FROM `Devices` where id = ?', [req.params.id],
            function (error, results, fields) {
                if (error) {
                    console.log(error);
                    res.send({ state: "An internal server error ocurred" }).status(500);
                }
                else {
                    if (results.length > 0) {
                        // We get the device
                        var currentDevice = results[0];
                        if ('name' in req.body)
                            currentDevice.name = req.body.name;
                        if ('description' in req.body)
                            currentDevice.description = req.body.description;
                        if ('type' in req.body) {
                            currentDevice.type = req.body.type;
                            currentDevice.state = 0; // We reset device state when updating type
                        }
                        if (deviceDataIsValid(currentDevice)) {
                            utils.query('UPDATE `Devices` set name = ?, description = ?, type = ?, state = ? where id= ?', [currentDevice.name, currentDevice.description, currentDevice.type, currentDevice.state, req.params.id],
                                function (error, results, fields) {
                                    if (error) {
                                        console.log(error);
                                        res.status(500).send({ state: "An internal server error ocurred while updating device" });
                                    }
                                    else {
                                        res.sendStatus(200);
                                    }
                                });
                        }
                        else
                            res.status(400).send({ state: "Device data is invalid" });;
                    }
                    else
                        res.sendStatus(404);
                }
            });
    }
    catch (e) {
        console.log(e);
        res.status(500).send({ state: "An internal server error ocurred" });
    }
});

app.listen(PORT, function (req, res) {
    console.log({ state: "NodeJS API running correctly" });
});

//=======[ End of file ]=======================================================
